R/data.table.R

```diff
@@ -1861,90 +1861,243 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
       }
       else { # rownames is a column number already
-        rownames = as.integer(rownames)
-        if (is.na(rownames) || rownames<1L || rownames>ncol(x))
+        rownames.index = as.integer(rownames)
+        if (is.na(rownames.index) || rownames.index<1L || rownames.index>ncol(x))
           stop("as.integer(rownames)==", rownames,
                " which is outside the column number range [1,ncol=", ncol(x), "].")
       }
     }
+    if (!is.null(rownames)) {
+      # Extract the rownames column, bringing it into memory if ff object
+      if (is.ff(x[[rownames.index]])) { 
+        rownames.value = x[[rownames.index]][]
+      } else {
+        rownames.value = x[[rownames.index]]
+      }
+      # Check rownames column is appropriate dimension
+      if (length(dim(rownames.value)) > 1L) {
+        stop("x[,", rownames, "] has multi-column type (such as a matrix column)",
+             " and cannot be used as rownames: dim(x[,", rownames, "])==", 
+             brackify(dim(rownames.value)))
+      }
+      # Warn if list column:
+      if (is.list(rownames.value)) {
+        warning("x[,", rownames, "] is a list column, which will be coerced to a",
+                "character vector when used as rownames")
+      }
+    }
   } else if (!is.null(rownames.value)) {
     if (length(rownames.value)!=nrow(x))
       stop("length(rownames.value)==", length(rownames.value),
            " but should be nrow(x)==", nrow(x))
   }
+  
+  # Create a new list X containing pointers to columns in x, so that any
+  # modifications (i.e. type coercion, dropping of rownames column) do 
+  # not change the input data.table. New copies in memory are only 
+  # created for columns that are modified (those coerced to a different
+  # type or class)
+  X = x
+  dm = dim(x)
+  cn = names(x)
+  class(X) = NULL
+  p = length(cn)
+  n = dm[1L]
+  
+  # If there is a rownames column, exctract, drop, and update the dimensions
   if (!is.null(rownames)) {
-    # extract that column and drop it.
-    rownames.value = x[[rownames]]
-    dm = dim(x) - 0:1
-    cn = names(x)[-rownames]
-    X = x[, .SD, .SDcols = cn]
-  } else {
-    dm = dim(x)
-    cn = names(x)
-    X = x
-  }
+    X[[rownames.index]] = NULL
+    dm = dm - 0:1
+    cn = cn[-rownames.index]
+    p = p - 1L
+  }
+  
+  # The maximum dimension size (row or column) for a matrix is 2^31-1 (or the Machine maximum integer)
+  # Check and error before doing computation/memory expensive column checks and coercion
+  if (p > .Machine$integer.max) 
+    error("Matrices with > ", .Machine$integer.max, " (.Machine$integer.max) columns are not supported")
+  if (n > INT_MAX) 
+    error("Matrices with > ", .Machine$integer.max, " (.Machine$integer.max) rows are not supported")
+  
+  # Check for any wide matrix like columns and unpack using as.data.table
+  if (length(which_wide_columns(x)) > 0L) {
+    X = as.data.table(X)
```

Comment:
Shouldn't we do this earlier?

E.g. the `p>.Machine$integer.max` test could in principle fail if `as.data.table` balloons `p` from < to >  `INT_MAX`